source/contributing/implementation/datamanipulator.rst

```diff
@@ -0,0 +1,467 @@
+    "5", "the maximum acceptable value", "bounded values"
+    "6", "the current value", "every subclass of ``BaseValue``, can be omitted"
+
+If no current value is specified, calling ``get()`` on the ``Value`` returns the default value.
+
+Copying and Serialization
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The two methods ``copy()`` and ``asImmutable()`` are not much work to implement. For both you just need to return
+a mutable or an immutable data manipulator respectively, containing the same data as the current instance.
+
+The method ``toContainer()`` is used for serialization purposes. Use a ``MemoryDataContainer`` as the result
+and apply to it the values stored within this instance. A ``DataContainer`` is basically a map mapping ``DataQuery``\ s
+to values. Since a ``Key`` always contains a corresponding ``DataQuery``, just use those by passing the ``Key`` directly.
+
+.. code-block:: java
+
+    public DataContainer toContainer() {
+        return new MemoryDataContainer()
+            .set(Keys.HEALTH, this.currentHealth)
+            .set(Keys.MAX_HEALTH, this.maximumHealth);
+    }
+
+registerGettersAndSetters()
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A ``DataManipulator`` also provides methods to get and set data using keys. The implementation for this is handled
+by ``AbstractData``, but we must tell it which data it can access and how. Therefore, in the
+``registerGettersAndSetters()`` method we need to do the following for each value:
+
+* register a ``Supplier`` to directly get the value
+* register a ``Consumer`` to directly set the value
+* register a ``Supplier<Value>`` to get the mutable ``Value``
+
+``Supplier`` and ``Consumer`` are functional interfaces, so Java 8 Lambdas can be used.
+
+.. code-block:: java
+
+ private void registerGettersAndSetters() {
+      registerFieldGetter(Keys.HEALTH, () -> SpongeHealthData.this.currentHealth);
+      registerFieldSetter(Keys.HEALTH, value -> SpongeHealthData.this.currentHealth = value);
+      registerKeyValue(Keys.HEALTH, SpongeHealthData.this::health);
+  }
+
+That's it. The ``DataManipulator`` should be done now.
+
+2. Implement the ImmutableDataManipulator
+=========================================
+
+Implementing the ``ImmutableDataManipulator`` is similar to implementing the mutable one.
+
+The only differences are:
+
+* The class name is formed by prefixing the mutable ``DataManipulator``\ s name with ``Immutable``
+* Inherit from ``ImmutableAbstractData`` instead
+* Instead of ``registerGettersAndSetters()``, the method is called ``registerGetters()``
+
+When creating ``ImmutableDataHolder``\ s or ``ImmutableValue``\ s, check if it makes sense to use the
+``ImmutableDataCachingUtil``. For example if you have ``WetData`` which contains nothing more than a boolean, it
+is more feasible to retain only two cached instances of ``ImmutableWetData`` - one for each possible value. For
+manipulators and values with many possible values (like ``SignData``) however, caching may prove too expensive.
```

Comment:
not may, but is.

To put it in perspective. It's cheap to cache predictable limited possibilities of values, but it's expensive to cache potentially limitless possible values.